samples/cellular/coap_client/src/main.c

```diff
@@ -220,41 +237,52 @@ static int wait(int timeout)
 	return 0;
 }
 
-int main(void)
+/**
+ * @brief Waits for the establishment of a network connection.
+ */
+static void wait_for_network(void)
+{
+	k_mutex_lock(&network_connected_lock, K_FOREVER);
```

Comment:
Have you considered making is_connected an atomic variable?
https://docs.zephyrproject.org/latest/kernel/services/other/atomic.html#defining-an-atomic-variable
Removing the need to protect the variable with the mutex

then you should be able to get a way with using a simple sema? This might reduce the complexity a bit.

Reply:
Keeping current solution, as discussed in person.